readthedocs/core/middleware.py

```diff
@@ -12,52 +12,71 @@
 log = logging.getLogger(__name__)
 
 LOG_TEMPLATE = u"(Middleware) {msg} [{host}{path}]"
+SUBDOMAIN_URLCONF = getattr(
+    settings,
+    'SUBDOMAIN_URLCONF',
+    'readthedocs.core.subdomain_urls'
+)
+SINGLE_VERSION_URLCONF = getattr(
+    settings,
+    'SINGLE_VERSION_URLCONF',
+    'readthedocs.core.single_version_urls'
+)
 
 
 class SubdomainMiddleware(object):
 
     def process_request(self, request):
+        if not getattr(settings, 'USE_SUBDOMAIN', False):
+            return None
+
         host = request.get_host().lower()
         path = request.get_full_path()
         log_kwargs = dict(host=host, path=path)
-        if settings.DEBUG:
-            log.debug(LOG_TEMPLATE.format(msg='DEBUG on, not processing middleware', **log_kwargs))
-            return None
+        public_domain = getattr(settings, 'PUBLIC_DOMAIN', None)
+        production_domain = settings.PRODUCTION_DOMAIN
+
+        if public_domain is None:
+            public_domain = production_domain
         if ':' in host:
             host = host.split(':')[0]
         domain_parts = host.split('.')
 
         # Serve subdomains - but don't depend on the production domain only having 2 parts
-        if len(domain_parts) == len(settings.PRODUCTION_DOMAIN.split('.')) + 1:
+        if len(domain_parts) == len(public_domain.split('.')) + 1:
             subdomain = domain_parts[0]
             is_www = subdomain.lower() == 'www'
             is_ssl = subdomain.lower() == 'ssl'
-            if not is_www and not is_ssl and settings.PRODUCTION_DOMAIN in host:
+            if not is_www and not is_ssl and public_domain in host:
                 request.subdomain = True
                 request.slug = subdomain
-                request.urlconf = 'readthedocs.core.subdomain_urls'
+                request.urlconf = SUBDOMAIN_URLCONF
                 return None
         # Serve CNAMEs
-        if settings.PRODUCTION_DOMAIN not in host and \
-           'localhost' not in host and \
-           'testserver' not in host:
+        if (public_domain not in host and
+                production_domain not in host and
+                'localhost' not in host and
+                'testserver' not in host):
```

Comment:
Nevermind. Let's do this later, I'll open a ticket. Moving settings to a constant file caused headaches with overriding settings in tests, and required mocking to work. There's probably a better pattern to be had here.